### PR TITLE
Split helix_core::find_root and helix_loader::find_local_config_dirs

### PR DESCRIPTION
**Pull request by rbino**
The documentation of find_root described the following priority for detecting a project root:
- Top-most folder containing a root marker in current git repository
- Git repository root if no marker detected
- Top-most folder containing a root marker if not git repository detected
- Current working directory as fallback

The commit contained in https://github.com/helix-editor/helix/pull/1249 extracted and changed the implementation of find_root in find_root_impl, actually reversing its result order (since that is the order that made sense for the local configuration merge, from innermost to outermost ancestors).

Since the two uses of find_root_impl have different requirements (and it's not a matter of reversing the order of results since, e.g., the top repository dir should be used by find_root only if there's not marker in other dirs), this PR splits the two implementations in two different specialized functions.

In doing so, find_root_impl is removed and the implementation is moved back in find_root, moving it closer to the documented behaviour thus making it easier to verify it's actually correct

### CODE REVIEW

**Review by kirawi — APPROVED** [review]
Nice catch.

File: helix-core/src/lib.rs [hunk 1]
```diff
@@ -47,9 +47,40 @@ pub fn find_first_non_whitespace_char(line: RopeSlice) -> Option<usize> {
...
+            top_marker = Some(ancestor);
+        }
+
+        // Don't go higher than repo if we're in one
+        if ancestor.join(".git").is_dir() {
+            // Top marker is repo root if not root marker is detected
+            if top_marker.is_none() {
+                top_marker = Some(ancestor);
+            }
+        }
```
Comment (the-mikedavis) [inline helix-core/src/lib.rs:79]: These two `if`s can be combined

Comment (rbino) [reply helix-core/src/lib.rs]: Actually a `break` was missing to stop the search as soon as we hit the git repo root

File: helix-core/src/lib.rs [hunk 2]
```diff
@@ -47,9 +47,41 @@ pub fn find_first_non_whitespace_char(line: RopeSlice) -> Option<usize> {
...
+            if top_marker.is_none() {
+                top_marker = Some(ancestor);
+            }
+            // Don't go higher than repo if we're in one
+            break;
+        }
+    }
+
+    // Return the found top marker or the current_dir as fallback
+    top_marker.map(|a| a.to_path_buf()).or(Some(current_dir))
```
Comment (the-mikedavis) [inline helix-core/src/lib.rs:84]: It looks like it's not possible for this Option to be returned as None now, so let's make this function return just a `std::path::PathBuf`

**Review by the-mikedavis — APPROVED** [review]

**Comment by archseer** [comment]
Nice catch! Thanks for a detailed writeup.

**Pull request merged**